### ISSUE TITLE
Apply input validator via ts decorator

### ISSUE THREAD

**Issue by tobiasdiez** [Code-Hex/graphql-codegen-typescript-validation-schema#190]
I thought of sharing a way to validate the input object via typescript decorators when implementing resolvers. Its designed to work with the typescript resolver plugin and uses the convention that mutations have to pass their arguments via a single `input` argument. Feel free to add this method in the generated code, or maybe add it as an example in the documentation. 

**Usage:**
```ts
class Mutation {
  @validateInput(SignupInputSchema)
  async signup(
    _root: Record<string, never>,
    { input: { email, password } }: MutationSignupArgs,
    context: Context
  ): Promise<SignupPayload> {
    // The input here is automatically valid to adhere to SignupInputSchema
  }
}
```

**Implementation:**
```ts
type ZodResolver<T extends ZodType<any, any, any>> = ResolverFn<
  any,
  any,
  any,
  { input: TypeOf<T> }
>

/**
 * Method decorator that validates the argument of the target function against the given schema.
 *
 * @export
 * @template T The type of the zod schema.
 * @param {T} arg The zod schema used for the validation.
 * @return {MethodDecorator} A {@link MethodDecorator}.
 */
export function validateInput<T extends AnyZodObject>(
  arg: T | (() => T)
): MethodDecorator<ZodResolver<T>> {
  return function (_target, _propertyKey, descriptor) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const originalMethod = descriptor.value!
    // @ts-expect-error: should be fine
    descriptor.value = function (root, { input }, context, info) {
      const schema = typeof arg === 'function' ? arg() : arg
      const result = schema.safeParse(input)

      if (result.success) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return originalMethod.call(
          this,
          root,
          { input: result.data },
          context,
          info
        )
      } else {
        return { problems: result.error.issues }
      }
    }
    return descriptor
  }
}
`̀ `

**Comment by rohit267** [comment]
this repo is not active

**Comment by Code-Hex** [comment]
Addressed in eef7987

**Comment by skizzo** [comment]
Somehow this isn't working for me in February 2023.

I'm using 
- @graphql-codegen/cli `5.0.2`
- @graphql-codegen/typescript `4.0.4`
- @graphql-codegen/typescript-resolvers `4.0.4`
- zod `3.22.4`
- typescript `4.8.4`

and this code, taken from both [the zod example](https://github.com/Code-Hex/graphql-codegen-typescript-validation-schema/tree/main/example/zod) and this very issue here:
```
import {AnyZodObject, TypeOf, ZodType} from "zod" // 3.22.4
import {ResolverFn} from "./GQL.types" // generated types

type ZodResolver<T extends ZodType<any, any, any>> = ResolverFn<any, any, any, {input: TypeOf<T>}>

/**
 * Method decorator that validates the argument of the target function against the given schema.
 *
 * @export
 * @template T The type of the zod schema.
 * @param {T} arg The zod schema used for the validation.
 * @return {MethodDecorator} A {@link MethodDecorator}.
 */
export function validateInput<T extends AnyZodObject>(arg: T | (() => T)): MethodDecorator<ZodResolver<T>> {
  return function (_target, _propertyKey, descriptor) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const originalMethod = descriptor.value!
    // @ts-expect-error: should be fine
    descriptor.value = function (root, {input}, context, info) {
      const schema = typeof arg === "function" ? arg() : arg
      const result = schema.safeParse(input)

      if (result.success) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return originalMethod.call(this, root, {input: result.data}, context, info)
      } else {
        return {problems: result.error.issues}
      }
    }
    return descriptor
  }
}

```

leads to the error 

## `Type 'MethodDecorator' is not generic. (typescript(2315)`:

<img width="773" alt="Screenshot 2024-02-13 at 14 48 35" src="https://github.com/Code-Hex/graphql-codegen-typescript-validation-schema/assets/2892572/522ebe83-2ae5-45ef-9d0e-c4e8165b931c">

### codegen.ts
```
import type {CodegenConfig} from "@graphql-codegen/cli"

const config: CodegenConfig = {
  schema: [
    //
    "./src/public/graphql/schema.discogs.graphql",
    "./src/public/graphql/schema._queries.graphql",
  ],
  generates: {
    "./src/_shared/to-app/types/GQL.types.ts": {
      plugins: ["typescript", "typescript-resolvers"],
      config: {
        useIndexSignature: true,
        typesPrefix: "TGQL",
        contextType: "./GraphQl.types#TGraphQlContext",
      },
    },
    "./src/_shared/to-app/zod/GQL.zod.ts": {
      plugins: ["typescript", "typescript-validation-schema"],
      config: {
        strictScalars: true,
        schema: "zod",
        withObjectType: true,
        enumsAsTypes: true,
        useTypeImports: true,
      },
    },
  },
}
export default config
```
Could somebody point me into the right direction here? Thanks.
